WIP: dojson: support enhanced marcxml links

ioannistsanaktsidis opened the pull request:
* Closes #485.

Signed-off-by: Ioannis Tsanaktsidis <ioannis.tsanaktsidis@cern.ch>

kaplun commented:
One regression test to go :+1:

kaplun commented:
Please, be more verbose in the commit log, and explain the changes (imagine github is down :smile: ) You can cut and paste all the details on MARC :)

ioannistsanaktsidis commented:
I addressed all the bullets from the issue. Concerning this one 773 $0 in HEP should go into into new book.recid field if BOOKCHAPTER I have implemented something in model.py. It seems to work for me but I am not about 100% abou the way I do it. :/

jmartinm commented:
@ioannistsanaktsidis There are still a couple of tests failing.

ioannistsanaktsidis commented:
Yes I know, I will check it today so as to complete it.

jmartinm commented on inspirehep/base/templates/format/record/Inspire_Default_HTML_brief.tpl:
```diff
@@ -51,11 +51,9 @@
                     {% endif %}
                   </b>
                 </h4>
-                {% if record.authors %}
```
Seems a bit unrelated to this PR, this if is not needed?

ioannistsanaktsidis replied:
I changed this and the general_macros file due to the changes I made in the collaboration field so as not to break it. Do you think we should move it to another PR?

jmartinm commented on inspirehep/dojson/hep/fields/bd70x75x.py:
```diff
@@ -57,7 +57,13 @@ def thesis_supervisor2marc(self, key, value):
-        return value.get('g')
+        recid = ''
+        if '0' in value:
+            recid = int(value.get('0'))
```
This will crash if there is no key `0` in value. You can do the `try...except` like above.

jmartinm commented on inspirehep/dojson/hep/fields/bd76x78x.py:
```diff
@@ -41,17 +42,20 @@ def publication_info(self, key, value):
-            if isinstance(value['0'], list):
-                recid = int(value['0'][0])
+        if '2' in value:
+            if isinstance(value['2'], list):
```
It's better to access all values with `value.get` to avoid crashing.

jmartinm commented on inspirehep/dojson/hep/model.py:
```diff
@@ -26,4 +26,19 @@
 hep = Overdo()
+
+
+@hep.over('book', '^773..')
```
Looking at the `dojson` source code I don't think this ensures that this code will always be run at the end (after the collections have been populated)

With a bit of Python magic, I have created this PR (https://github.com/inspirehep/inspire-next/pull/666) to show another way that would ensure we execute the code at the end.

@kaplun @jalavik what do you think? This is in order to allow for some code to execute after dojson is finished (to ensure all the keys are present)

ioannistsanaktsidis replied:
@jmartinm  you are right, thats what I wanted to discuss also. This thing that I did is not running for every record. Sometimes it is missing.

jalavik replied:
Yeah, the order is not guaranteed. In the past we would add the extra stuff directly in the function (in this case collections), but it is not always nice as collections is now shared between all dojsons in `base.py`.

@jmartinm Another possibly less "magic" way could be to use inheritence to subclass the `Overdo` class in the hep model and there override the `do()` function to call the super version and then do our code.

jmartinm replied:
Good point @jalavik Actually did not read this comment in time and merged https://github.com/inspirehep/inspire-next/pull/666 after a slight modification that allows to keep using `hep.do`

The effect is the same, so at least somebody can learn how to inject a function into a class instance :)

ioannistsanaktsidis replied:
I guess this someone is me :P :D
On Dec 15, 2015 6:24 PM, "Javier Martin Montull" <notifications@github.com>
wrote:

> In inspirehep/dojson/hep/model.py
> <https://github.com/inspirehep/inspire-next/pull/636#discussion_r47667895>
> :
>
> > @@ -26,4 +26,19 @@
> >  from dojson import utils
> >
> >  hep = Overdo()
> > +
> > +
> > +@hep.over('book', '^773..')
>
> Good point @jalavik <https://github.com/jalavik> Actually did not read
> this comment in time and merged #666
> <https://github.com/inspirehep/inspire-next/pull/666> after a slight
> modification that allows to keep using hep.do
>
> The effect is the same, so at least somebody can learn how to inject a
> function into a class instance :)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/inspirehep/inspire-next/pull/636/files#r47667895>.
>

ioannistsanaktsidis commented:
All tests are passing. We can wait for @jmartinm  PR to be merged and then we can merge this one.

jmartinm commented:
Continues here https://github.com/inspirehep/inspire-next/pull/674
